bundle/src/test/java/com/adobe/acs/commons/mcp/DialogResourceProviderFactoryTest.java

```diff
@@ -100,23 +103,63 @@ public void testResourceResolution() {
     }
 
     @Test
-    public void testDynamicRegistration() {
-        assertEquals("Should have 3 services registered", 3, dialogProvider.getActiveProviders().size());
+    public void testDeactivation() {
+        assertTrue("Should resolve model one dialog", resourceExists("/apps/test/model1/cq:dialog"));
+        providers.get("com.adobe.acs.commons.mcp.model.SimpleModelOne").doDeactivate();
+        assertFalse("Should resolve model one dialog", resourceExists("/apps/test/model1/cq:dialog"));
+    }
 
-        AdapterFactory adapterFactory = Mockito.mock(AdapterFactory.class);
-        Map<String,Object> properties = new HashedMap<>();
-        properties.put("models.adapter.implementationClass", "com.adobe.acs.commons.mcp.model.SimpleModelFour");
-        dialogProvider.bind(adapterFactory, properties);
+    @Test
+    public void testFactoryDoesNothing() {
+        // The dialog provider factory is dead but still exists for backward compatiblity until 5.0
+        // Assert that it does nothing.
+        DialogResourceProviderFactory factory = new DialogResourceProviderFactoryImpl();
+        assertNull(factory.getActiveProviders());
 
-        assertEquals("Should have 4 services registered", 4, dialogProvider.getActiveProviders().size());
-        assertTrue("Should resolve model four dialog", resourceExists("/apps/test/model4/cq:dialog"));
+        // These would normally throw an exception on null data if they did anything.
+        factory.registerClass((Class) null);
+        factory.registerClass((String) null);
+        factory.unregisterClass((Class) null);
+        factory.unregisterClass((String) null);
+    }
 
-        dialogProvider.unbind(adapterFactory, properties);
-        assertEquals("Should have 3 services registered", 3, dialogProvider.getActiveProviders().size());
-        assertFalse("Should NOT resolve model four dialog", resourceExists("/apps/test/model4/cq:dialog"));
+    @Test
+    public void testAnnotationProvider() {
```

Comment:
Negative test cases? 
test cases when the annotation is added to an invalid class definition? Or one that has `getResourceType` and/or a field `resourceType` and/or the Resource adaptable annotation?

Reply:
Negative discrimination, as in the annotation processor sees a class with the annotation but it lacks a resource type so it decides not to render an OSGi service.  This is because rendering an OSGi service with no resource type will result in an error later when that service tries to start and cannot figure out where in the JCR to latch itself.

Reply:
So there's some positive and negative discrimination cases in the unit test to ensure that logic works -- in addition I also wrote something that calls the annotation processor directly so I could improve the code coverage.  Honestly I could just keep that google compiler test instead of the discrimination tests as they do the same thing, but the first set confirm that the annotation processor fired off in Maven prior to testing, the other runs the processor more directly within the test.

Reply:
However the google compiler library only lets me check if a class was written. Funny enough it doesn't have a negative assert rule to assert that a class was NOT written.  Oh well.

Reply:
There are cases where the annotation is used in a base class that is itself not a model and does not have a resource type.  Because the annotation supports inheritance I allow for it, such that the base class does not get a resource provider but its fully-implemented subclasses will.

Reply:
This is also what allows all these code changes to be backwards compatible with no breaking changes from the previous code.  I would encourage people to use the annotation directly on a model class as it is clearer what it is doing, but realistically there's some flexibility if people want to assert that on a base class and keep their model classes super-lean.

Reply:
Ok, that helps. I was trying to figure out why there weren't negative validations, but this clarifies. you log and move on, and the onus is on the consumer to figure out why they aren't getting their services (e.g. they didn't meet the api spec.